File: 0310-minimum-height-trees/0310-minimum-height-trees.py
class Solution:
    def findMinHeightTrees(self, n: int, edges: List[List[int]]) -> List[int]:
        tree = defaultdict(list)
        heights = defaultdict(int)
        indegree = [0 for i in range(n)]
        visited = set()
        
        if len(edges) == 0:
            return [0]
        
        for source, des in edges:
            tree[source].append(des)
            tree[des].append(source)
          
        queue = deque()
        for i in range(n):
            indegree[i] = len(tree[i])
            
            if indegree[i] == 1:
                visited.add(i)
                queue.append(i)
            
        while queue:
            length = len(queue)
            if max(indegree) == 1:
                return queue
            
            for _ in range(length):
                node = queue.popleft()

                for neighbor in tree[node]:
                    indegree[neighbor] -= 1
                   
                    if neighbor not in visited:
                        if indegree[neighbor] <= 1:
                            visited.add(neighbor)
                            queue.append(neighbor)
           
        return []              
                    

File: 0310-minimum-height-trees/0310-minimum-height-trees.js
/**
 * @param {number} n
 * @param {number[][]} edges
 * @return {number[]}
 */
var findMinHeightTrees = function(n, edges) {
    const tree = {};
        const heights = {};
        const indegree = Array(n).fill(0);
        const visited = new Set();

        if (edges.length === 0) {
            return [0];
        }

        for (const [source, des] of edges) {
            if (!tree[source]) {
                tree[source] = [];
            }
            if (!tree[des]) {
                tree[des] = [];
            }
            tree[source].push(des);
            tree[des].push(source);
        }

        const queue = [];
        for (let i = 0; i < n; i++) {
            indegree[i] = tree[i] ? tree[i].length : 0;

            if (indegree[i] === 1) {
                visited.add(i);
                queue.push(i);
            }
        }

        while (queue.length > 0) {
            const length = queue.length;

            if (Math.max(...indegree) === 1) {
                return queue;
            }

            for (let i = 0; i < length; i++) {
                const node = queue.shift();

                for (const neighbor of tree[node]) {
                    indegree[neighbor] -= 1;

                    if (!visited.has(neighbor)) {
                        if (indegree[neighbor] <= 1) {
                            visited.add(neighbor);
                            queue.push(neighbor);
                        }
                    }
                }
            }
        }

        return [];
};